client/session.go

```diff
@@ -501,80 +542,113 @@ func (s *session) FetchAll(namespace string, ids []string, startInclusive, endEx
 
 	fetchBatchOpsByHostIdx = s.fetchBatchOpArrayArrayPool.Get()
 
-	// Defer cleanup
-	defer func() {
-		// Return fetch ops to pool
-		for _, ops := range fetchBatchOpsByHostIdx {
-			for _, op := range ops {
-				s.fetchBatchOpPool.Put(op)
-			}
-		}
-
-		// Return fetch ops array array to pool
-		s.fetchBatchOpArrayArrayPool.Put(fetchBatchOpsByHostIdx)
-	}()
-
 	s.RLock()
-	majority = s.majority
+	majority = int32(s.majority)
 	for idx := range ids {
 		idx := idx
 
 		var (
+			wgIsDone int32
 			results  []encoding.Iterator
-			enqueued int
+			enqueued int32
 			pending  int32
 			success  int32
 			firstErr error
 			errs     int32
 		)
 
 		wg.Add(1)
+		allCompletionFn := func() {
+			var reportErr error
+			errsLen := atomic.LoadInt32(&errs)
+			if errsLen > 0 {
+				resultErrLock.RLock()
+				reportErr = firstErr
+				resultErrLock.RUnlock()
+			}
+			if err := s.consistencyResult(majority, enqueued, errsLen, reportErr); err != nil {
+				resultErrLock.Lock()
+				if resultErr == nil {
+					resultErr = err
+				}
+				resultErrs++
+				resultErrLock.Unlock()
+			} else {
+				// This is the final result, its safe to load "success" now without atomic operation
+				iter := s.seriesIteratorPool.Get()
+				iter.Reset(ids[idx], startInclusive, endExclusive, results[:atomic.LoadInt32(&success)])
```

Comment:
Do we even return this iter to the seriesIteratorPool? Can't see that in this diff.